File: backend/app/api/routers/vector.py
from fastapi import APIRouter, Depends, HTTPException, status
from sqlalchemy.ext.asyncio import AsyncSession

from ...core.dependencies import get_current_user
from ...db.session import get_session
from ...models import User as UserInDB
from ...schemas.admin import (
    VectorChapterBatchRequest,
    VectorChapterDetailResponse,
    VectorChapterListResponse,
    VectorOperationResult,
    VectorProjectListResponse,
    VectorRetrievalTestRequest,
    VectorRetrievalTestResponse,
)
from ...services.vector_management_service import VectorManagementService

router = APIRouter(prefix="/api/vector", tags=["Vector"])


def get_vector_service(session: AsyncSession = Depends(get_session)) -> VectorManagementService:
    return VectorManagementService(session)


@router.get("/projects", response_model=VectorProjectListResponse)
async def list_projects(
    service: VectorManagementService = Depends(get_vector_service),
    current_user: UserInDB = Depends(get_current_user),
) -> VectorProjectListResponse:
    return await service.list_projects(user_id=current_user.id)


@router.get("/projects/{project_id}/chapters", response_model=VectorChapterListResponse)
async def list_project_chapters(
    project_id: str,
    service: VectorManagementService = Depends(get_vector_service),
    current_user: UserInDB = Depends(get_current_user),
) -> VectorChapterListResponse:
    return await service.list_project_chapters(project_id, user_id=current_user.id)


@router.get(
    "/projects/{project_id}/chapters/{chapter_number}",
    response_model=VectorChapterDetailResponse,
)
async def get_chapter_detail(
    project_id: str,
    chapter_number: int,
    service: VectorManagementService = Depends(get_vector_service),
    current_user: UserInDB = Depends(get_current_user),
) -> VectorChapterDetailResponse:
    return await service.get_chapter_detail(project_id, chapter_number, user_id=current_user.id)


@router.post(
    "/projects/{project_id}/chapters/reingest",
    response_model=VectorOperationResult,
)
async def reingest_chapters(
    project_id: str,
    payload: VectorChapterBatchRequest,
    service: VectorManagementService = Depends(get_vector_service),
    current_user: UserInDB = Depends(get_current_user),
) -> VectorOperationResult:
    return await service.reingest_chapters(project_id, payload.chapter_numbers, user_id=current_user.id)


@router.post(
    "/projects/{project_id}/chapters/delete",
    response_model=VectorOperationResult,
)
async def delete_chapters(
    project_id: str,
    payload: VectorChapterBatchRequest,
    service: VectorManagementService = Depends(get_vector_service),
    current_user: UserInDB = Depends(get_current_user),
) -> VectorOperationResult:
    return await service.delete_chapters(project_id, payload.chapter_numbers, user_id=current_user.id)


@router.post(
    "/projects/{project_id}/retrieval-test",
    response_model=VectorRetrievalTestResponse,
)
async def test_retrieval(
    project_id: str,
    payload: VectorRetrievalTestRequest,
    service: VectorManagementService = Depends(get_vector_service),
    current_user: UserInDB = Depends(get_current_user),
) -> VectorRetrievalTestResponse:
    return await service.test_retrieval(project_id, payload, user_id=current_user.id)

File: frontend/src/api/vector.ts
import router from '@/router'
import { useAuthStore } from '@/stores/auth'

const API_BASE_URL = import.meta.env.MODE === 'production' ? '' : 'http://127.0.0.1:8000'
const VECTOR_PREFIX = '/api/vector'

const request = async (path: string, options: RequestInit = {}) => {
  const authStore = useAuthStore()
  const headers = new Headers({
    'Content-Type': 'application/json',
    ...options.headers
  })

  if (authStore.isAuthenticated && authStore.token) {
    headers.set('Authorization', `Bearer ${authStore.token}`)
  }

  const response = await fetch(`${API_BASE_URL}${VECTOR_PREFIX}${path}`, {
    ...options,
    headers
  })

  if (response.status === 401) {
    authStore.logout()
    router.push('/login')
    throw new Error('会话已过期，请重新登录')
  }

  if (!response.ok) {
    const errorData = await response.json().catch(() => ({}))
    throw new Error(errorData.detail || `请求失败，状态码: ${response.status}`)
  }

  if (response.status === 204) {
    return
  }

  return response.json()
}

export interface VectorProjectSummary {
  project_id: string
  title: string
  total_chapters: number
  ingested_chapters: number
  partial_chapters: number
  missing_chapters: number
  stale_chapters: number
  last_ingested_at?: string | null
  has_vectors: boolean
}

export interface VectorProjectListResponse {
  projects: VectorProjectSummary[]
  vector_db_size_bytes?: number | null
}

export interface VectorChapterSummary {
  chapter_number: number
  title: string
  status: 'ingested' | 'partial' | 'missing' | 'stale'
  chunk_count: number
  summary_count: number
  last_ingested_at?: string | null
  confirmed: boolean
  needs_refresh: boolean
  updated_at?: string | null
  word_count?: number | null
}

export interface VectorChapterTotals {
  total: number
  ingested: number
  partial: number
  missing: number
  stale: number
}

export interface VectorChapterListResponse {
  project_id: string
  totals: VectorChapterTotals
  chapters: VectorChapterSummary[]
}

export interface VectorChunkDetail {
  chunk_index: number
  chapter_title?: string | null
  content: string
  embedding_dim: number
  metadata: Record<string, any>
  created_at?: string | null
}

export interface VectorSummaryDetail {
  title: string
  summary: string
  embedding_dim: number
  created_at?: string | null
}

export interface VectorChapterDetailResponse {
  project_id: string
  chapter_number: number
  chunks: VectorChunkDetail[]
  summary?: VectorSummaryDetail | null
}

export interface VectorOperationResult {
  processed: number
  skipped: number
  failed: number
  message?: string | null
}

export interface VectorRetrievalTestRequest {
  query: string
  top_k_chunks?: number
  top_k_summaries?: number
}

export interface VectorRetrievalChunk {
  chapter_number: number
  chunk_index?: number | null
  chapter_title?: string | null
  content: string
  score: number
}

export interface VectorRetrievalSummary {
  chapter_number: number
  title: string
  summary: string
  score: number
}

export interface VectorRetrievalTestResponse {
  query: string
  chunks: VectorRetrievalChunk[]
  summaries: VectorRetrievalSummary[]
}

export class VectorAPI {
  static getProjects(): Promise<VectorProjectListResponse> {
    return request('/projects')
  }

  static getProjectChapters(projectId: string): Promise<VectorChapterListResponse> {
    return request(`/projects/${projectId}/chapters`)
  }

  static getChapterDetail(projectId: string, chapterNumber: number): Promise<VectorChapterDetailResponse> {
    return request(`/projects/${projectId}/chapters/${chapterNumber}`)
  }

  static reingest(projectId: string, chapterNumbers: number[]): Promise<VectorOperationResult> {
    return request(`/projects/${projectId}/chapters/reingest`, {
      method: 'POST',
      body: JSON.stringify({ chapter_numbers: chapterNumbers })
    })
  }

  static delete(projectId: string, chapterNumbers: number[]): Promise<VectorOperationResult> {
    return request(`/projects/${projectId}/chapters/delete`, {
      method: 'POST',
      body: JSON.stringify({ chapter_numbers: chapterNumbers })
    })
  }

  static testRetrieval(projectId: string, payload: VectorRetrievalTestRequest): Promise<VectorRetrievalTestResponse> {
    return request(`/projects/${projectId}/retrieval-test`, {
      method: 'POST',
      body: JSON.stringify(payload)
    })
  }
}
